neqo-crypto/src/agent.rs

```diff
@@ -50,6 +50,110 @@ use crate::{
+    /// though the encoding is not frequently used
+    /// Enable decoding field is used to signal to the implementation
+    /// to use the encoding as well
+    const ENABLE_ENCODING: bool = false;
+
+    /// Certificate Compression encoding function
+    /// If the implementation is not provided, we only copy the data
+    /// NB: If `ENABLE_ENCODING` is not set, the function pointer provided to NSS will be null
+    #[expect(clippy::must_use_candidate, reason = "Encoders are optional")]
+    fn encode(data: &[u8]) -> Vec<u8> {
+        data.to_vec()
+    }
+
+    /// Certificate Compression decoding function
+    #[must_use]
+    fn decode(data: &[u8]) -> Vec<u8>;
+}
+
+/// The trait is responsible for calling `SafeCertificateCompression` encoding and decoding
+/// functions using the NSS types
+unsafe impl<T: SafeCertificateCompression> UnsafeCertCompression for T {
+    unsafe extern "C" fn decode_callback(
+        input: *const ssl::SECItem,
+        output: *mut ::std::os::raw::c_uchar,
+        output_len: usize,
+        used_len: *mut usize,
+    ) -> ssl::SECStatus {
+        unsafe {
+            match std::ptr::NonNull::new(input.cast_mut()) {
+                None => ssl::SECFailure,
+                Some(input) => {
+                    if input.as_ref().data.is_null() || input.as_ref().len == 0 {
+                        return ssl::SECFailure;
+                    }
+
+                    let bytes_to_decode_ptr =
+                        null_safe_slice(input.as_ref().data, input.as_ref().len);
+                    let decoded_bytes = T::decode(bytes_to_decode_ptr);
+                    if decoded_bytes.len() > output_len {
+                        return ssl::SECFailure;
+                    }
+
+                    std::ptr::copy_nonoverlapping(decoded_bytes.as_ptr(), output, output_len);
+                    *used_len = decoded_bytes.len();
+                    ssl::SECSuccess
+                }
+            }
+        }
+    }
+
+    unsafe extern "C" fn encode_callback(
+        input: *const ssl::SECItem,
+        output: *mut ssl::SECItem,
+    ) -> ssl::SECStatus {
+        unsafe {
+            match std::ptr::NonNull::new(input.cast_mut()) {
+                None => ssl::SECFailure,
+                Some(input) => {
+                    if input.as_ref().data.is_null() || input.as_ref().len == 0 {
+                        return ssl::SECFailure;
+                    }
+
+                    let bytes_to_encode = null_safe_slice(input.as_ref().data, input.as_ref().len);
+                    let encoded_bytes = T::encode(bytes_to_encode);
+
+                    p11::SECITEM_MakeItem(
+                        null_mut(),
+                        // p11::SECItem is the same as ssl::SECItem
+                        output.cast::<p11::SECItemStr>(),
+                        encoded_bytes.as_ptr(),
+                        encoded_bytes.len().try_into().unwrap(),
```

Comment:
I think that the general preference in this crate is to avoid `unwrap()`; you can instead do:

```rust
let Some(encoded_len) = c_uint::try_from(encoded_bytes.len()) else { return ssl::SECFailure; };
```

Reply:
Done!